File: 프로그래머스/1/12977. 소수 만들기/소수 만들기.cpp
#include <vector>
#include <iostream>

using namespace std;

bool isPrime(int sum) {
    for(int i = 2; i * i <= sum; i++) {
        if(sum % i == 0) {
            return false;
        }
    }
    return true;
}

int solution(vector<int> nums) {
    int answer = 0;

    for(int i = 0; i < nums.size(); i++) {
        for(int j = i + 1; j < nums.size(); j++) {
            for(int k = j + 1; k < nums.size(); k++) {
                if(isPrime(nums[i] + nums[j] + nums[k])) {
                    answer++;
                }
            }
        }
    }

    return answer;
}

// [ 소수 판별 알고리즘 ]

// 1. 소수인지 확인하기 위해 약수 존재 여부를 검사한다.
//    - 1과 자기 자신을 제외한 약수가 하나라도 있으면 소수가 아니다.

// 2. 약수 검사는 n의 제곱근(√n)까지만 하면 된다.
//    - 모든 약수는 짝이 있으므로, 한 쪽은 반드시 √n 이하에 존재하기 때문.

// 예) n = 36
//    1 * 36
//    2 * 18
//    3 * 12
//    4 * 9
//    6 * 6
//    → 작은 약수들은 모두 √36 = 6 이하이므로
//      2 ~ 6까지만 검사해도 소수가 아님을 알 수 있다.


        

File: 프로그래머스/1/12977. 소수 만들기/소수 만들기.java
class Solution {
    public int solution(int[] nums) {
        int answer = 0;

        for(int i = 0; i < nums.length; i++){
            for(int j = i + 1; j < nums.length; j++){
                for(int k = j + 1; k < nums.length; k++){
                    if(isPrime(nums[i]+nums[j]+nums[k])){
                        answer++;
                    }
                }
            }
        }
        return answer;
    }

    // 소수와 약수의 관계: 소수는 약수의 개수가 1과 본인 밖에 없기 때문에 약수 개수 판별 알고리즘 사용 가능
    private boolean isPrime(int num) {
        // 만약 num이 2인 경우 소수이므로 true 필요, 하지만 세 값을 더한 수는 항상 1보다 큼
        for(int i = 2; i * i <= num; i++) {
            if(num % i == 0) {
                return false;
            }
        }

        return true;
    }
}